File: BAEKJOON/1744 - 수 묶기/수 묶기.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// 최대 합을 구하는 함수
int max_sequence_sum(vector<int>& nums) {
    vector<int> positive;  // 양수를 저장하는 리스트
    vector<int> negative;  // 음수와 0을 저장하는 리스트
    int result = 0;  // 최종 결과값을 저장하는 변수

    // 주어진 수열을 순회하며 분류
    for (int num : nums) {
        if (num > 1) {  
            positive.push_back(num);  // 1보다 큰 양수는 positive 리스트에 저장
        } else if (num == 1) {
            result += 1;  // 1은 바로 더하는 것이 유리
        } else {
            negative.push_back(num);  // 음수와 0은 negative 리스트에 저장
        }
    }

    // 양수는 내림차순으로 정렬 (큰 것부터 묶기 위해)
    sort(positive.begin(), positive.end(), greater<int>());
    // 음수는 오름차순으로 정렬 (작은 것부터 묶기 위해)
    sort(negative.begin(), negative.end());

    // 양수 묶기
    int i = 0;
    while (i + 1 < positive.size()) {  // 인덱스 초과 방지
        result += positive[i] * positive[i + 1];  // 두 개씩 묶어 곱한 결과를 더함
        i += 2;  // 두 개씩 묶었으므로 인덱스를 2 증가
    }
    // 묶지 못하고 남은 양수 처리
    if (i < positive.size()) {
        result += positive[i];  // 남은 양수는 그대로 더함
    }

    // 음수 묶기
    i = 0;
    while (i + 1 < negative.size()) {  // 인덱스 초과 방지
        result += negative[i] * negative[i + 1];  // 두 개씩 묶어 곱한 결과를 더함
        i += 2;  // 두 개씩 묶었으므로 인덱스를 2 증가
    }
    // 묶지 못하고 남은 음수 처리
    if (i < negative.size()) {
        if (find(nums.begin(), nums.end(), 0) != nums.end()) {
            result += 0;  // 0이 있는 경우 음수와 묶어 상쇄
        } else {
            result += negative[i];  // 0이 없으면 음수를 그대로 더함
        }
    }

    return result;  // 최종 결과 반환
}

int main() {
    int n;
    cin >> n;  // 수열의 크기 입력
    vector<int> nums(n);  // 수열을 저장할 벡터
    for (int i = 0; i < n; i++) {
        cin >> nums[i];  // 수열의 각 수 입력
    }
    cout << max_sequence_sum(nums) << endl;  // 결과 출력
    return 0;
}

File: BAEKJOON/1744 - 수 묶기/수 묶기.py
def max_sequence_sum(nums):
    positive = []  # 양수를 저장할 리스트
    negative = []  # 음수와 0을 저장할 리스트
    result = 0  # 최종 결과를 저장할 변수

    for num in nums:  # 입력 받은 숫자들을 하나씩 확인
        if num > 1:  # 1보다 큰 양수는 positive 리스트에 추가
            positive.append(num)
        elif num == 1:  # 1은 묶지 않고 결과에 바로 더함
            result += 1
        else:  # 0과 음수는 negative 리스트에 추가
            negative.append(num)

    positive.sort(reverse=True)  # 양수는 큰 것부터 곱하기 위해 내림차순 정렬
    negative.sort()  # 음수는 작은 것부터 곱하기 위해 오름차순 정렬

    i = 0  # 양수 리스트의 인덱스를 관리할 변수
    while i < len(positive) - 1:  # 두 개씩 묶어서 곱하기
        result += (
            positive[i] * positive[i + 1]
        )  # 양수 두 개를 묶어 곱한 값을 결과에 추가
        i += 2  # 두 개씩 묶었으므로 인덱스를 2 증가시킴
    if i < len(positive):  # 묶지 못하고 남은 양수가 있으면 그대로 더하기
        result += positive[i]

    i = 0  # 음수 리스트의 인덱스를 관리할 변수
    while i < len(negative) - 1:  # 두 개씩 묶어서 곱하기
        result += (
            negative[i] * negative[i + 1]
        )  # 음수 두 개를 묶어 곱한 값을 결과에 추가
        i += 2  # 두 개씩 묶었으므로 인덱스를 2 증가시킴
    if i < len(negative):  # 묶지 못하고 남은 음수가 있으면
        if 0 in nums:  # 0이 있을 경우, 음수와 0을 묶어 결과에 0을 더함 (음수 상쇄)
            result += 0
        else:  # 0이 없으면 남은 음수를 결과에 그대로 더함
            result += negative[i]

    return result  # 최종 결과 반환


n = int(input())  # 수열의 크기 입력
nums = [int(input()) for _ in range(n)]  # 수열의 각 수를 입력 받아 리스트에 저장
print(max_sequence_sum(nums))  # 최종 결과 출력
